src/app/InteractionModelEngine.cpp

```diff
@@ -764,34 +860,131 @@ CHIP_ERROR InteractionModelEngine::CanEstablishReadTransaction(const ReadHandler
+    const size_t attributePathCap = allowUnlimited ? SIZE_MAX : GetPathPoolCapacityForReads();
+    const size_t eventPathCap     = allowUnlimited ? SIZE_MAX : GetPathPoolCapacityForReads();
+    const size_t readHandlerCap   = allowUnlimited ? SIZE_MAX : GetReadHandlerPoolCapacityForReads();
+
+    size_t usedAttributePaths = 0;
+    size_t usedEventPaths     = 0;
+    size_t usedReadHandlers   = 0;
 
-    // It is safe to use & here since this function will be called on current stack.
-    if (apReadHandler->GetAttributePathCount() > kReservedPathsPerReadRequest ||
-        apReadHandler->GetEventPathCount() > kReservedPathsPerReadRequest ||
-        apReadHandler->GetDataVersionFilterCount() > kReservedPathsPerReadRequest)
+    auto countResourceUsage = [&]() {
+        usedAttributePaths = 0;
+        usedEventPaths     = 0;
+        usedReadHandlers   = 0;
+        mReadHandlers.ForEachActiveObject([&](auto * handler) {
+            if (!handler->IsType(ReadHandler::InteractionType::Read))
+            {
+                return Loop::Continue;
+            }
+            usedAttributePaths += handler->GetAttributePathCount();
+            usedEventPaths += handler->GetEventPathCount();
+            usedReadHandlers++;
+            return Loop::Continue;
+        });
+    };
+
+    countResourceUsage();
+
+    if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+        usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
     {
-        // Doesn't matter how many other reads are in progress; we are not
-        // going to handle this request.
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(PathsExhausted));
+        // We have enough resources, then we serve the requests in a best-effort manner.
+        return Protocols::InteractionModel::Status::Success;
     }
 
-    mReadHandlers.ForEachActiveObject([&](ReadHandler * handler) {
-        if (handler->GetAccessingFabricIndex() == currentFabricIndex && handler->IsType(ReadHandler::InteractionType::Read))
+    if ((aRequestedAttributePathCount > kMinSupportedPathsPerReadRequest &&
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > kMinSupportedPathsPerReadRequest && usedEventPaths + aRequestedEventPathCount > eventPathCap))
+    {
+        // We cannot offer enough resources, and the read transaction is requesting more than the spec limit.
+        return Protocols::InteractionModel::Status::PathsExhausted;
+    }
+
+    // If we have commissioned CHIP_CONFIG_MAX_FABRICS already, and this transaction don't have an associated fabric index, reject
+    // the request if we don't have sufficient resources for this request.
+    if (mpFabricTable->FabricCount() == GetConfigMaxFabrics() && aFabricIndex == kUndefinedFabricIndex)
+    {
+        return Protocols::InteractionModel::Status::ResourceExhausted;
```

Comment:
Why is this not BUSY?

In particular, can we ever end up in a sitation where a PASE read is permanently blocked?

Reply:
It might starve (when the fabric table is full), but the resource are guaranteed if the fabric table is not full, test case 5,6,7,8 is for these situations.

Reply:
This is the case that the fabric table is full, and we did not have any resources reserved for the extra PASE session. So ResourceExhausted makes sense to me -- we cannot guarantee any resource for this session, even if it retried later, it might be evicted since we must serve the transactions from a PASE session.